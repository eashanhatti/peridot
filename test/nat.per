define Nat
  : Type
  = Record(
      tag : Bool,
      data :
        if(tag) {
          Record()
        } else {
          Nat
        }
    );

define zero
  : Nat
  = record(
      tag = true,
      data = record()
    );

define succ
  : Function(Nat) -> Nat
  = function(n) =>
      record(
        tag = false,
        data = n
      );

define add
  : Function(Nat, Nat) -> Nat
  = function(n, m) =>
      if(n.tag){
        m
      } else {
        add(n.data, m)
      };
