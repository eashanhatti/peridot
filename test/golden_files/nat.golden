((fromList [], fromList [], fromList []),
 Declare
   Obj
   (Rigid (NameObjIntro (Id {unId = 0})))
   (Rigid (TypeType Obj))
   (Define
      (Rigid (NameObjIntro (Id {unId = 0})))
      (RecType
         (fromList
            [(Field {unField = "tag"}, Rigid TwoType),
             (Field {unField = "data"},
              TwoElim
                (LocalVar (Index {unIndex = 0}))
                (RecType (fromList []))
                (GlobalVar (Id {unId = 0})))]))
      (Declare
         Obj
         (Rigid (NameObjIntro (Id {unId = 1})))
         (GlobalVar (Id {unId = 0}))
         (Declare
            Obj
            (Rigid (NameObjIntro (Id {unId = 2})))
            (ObjFunType
               Explicit
               (GlobalVar (Id {unId = 0}))
               (GlobalVar (Id {unId = 0})))
            (Declare
               Obj
               (Rigid (NameObjIntro (Id {unId = 3})))
               (ObjFunType
                  Explicit
                  (GlobalVar (Id {unId = 0}))
                  (ObjFunType
                     Explicit
                     (GlobalVar (Id {unId = 0}))
                     (GlobalVar (Id {unId = 0}))))
               (Define
                  (Rigid (NameObjIntro (Id {unId = 1})))
                  (RecIntro
                     (fromList
                        [(Field {unField = "tag"}, Rigid TwoIntro0),
                         (Field {unField = "data"}, RecIntro (fromList []))]))
                  (Define
                     (Rigid (NameObjIntro (Id {unId = 2})))
                     (ObjFunIntro
                        (RecIntro
                           (fromList
                              [(Field {unField = "tag"}, Rigid TwoIntro1),
                               (Field {unField = "data"},
                                LocalVar (Index {unIndex = 1}))])))
                     (Define
                        (Rigid (NameObjIntro (Id {unId = 3})))
                        (ObjFunIntro
                           (ObjFunIntro
                              (TwoElim
                                 (RecElim
                                    (LocalVar (Index {unIndex = 1}))
                                    (Field {unField = "tag"}))
                                 (LocalVar (Index {unIndex = 0}))
                                 (ObjFunElim
                                    (ObjFunElim
                                       (GlobalVar (Id {unId = 3}))
                                       (RecElim
                                          (LocalVar (Index {unIndex = 1}))
                                          (Field {unField = "data"})))
                                    (LocalVar (Index {unIndex = 0}))))))
                        (Rigid (TypeType Obj))))))))))
